Add support for ALTER VIEW RENAME TO statement

dmariamgeorge opened the pull request:
## Description
Add support for SQL statement ALTER VIEW <view_name> RENAME TO <new_view_name>. 

## Motivation and Context

Addresses issue: https://github.com/prestodb/presto/issues/23748

## Impact

Support SQL statement ALTER VIEW <view_name> RENAME TO <new_view_name> in Iceberg connector.

## Test Plan

- Modified `testTableRules()` in `TestFileBasedAccessControl.java` to include check for rename view. 
- Added test `testRenameView()` in `TestSqlParser.java`.
- In memory connector, added tests `testRenameView()` and `testRenameViewIfNotExists()` in `TestMemorySmoke.java`.
- Modified test `testViews()` in `TestMemoryMetadata.java` to include check for rename view. 
- In Iceberg connector, added tests `testRenameView()` and `testRenameViewIfNotExists()` in `TestIcebergMetadataListing.java`. 

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==
General Changes
* Add support for ``ALTER VIEW RENAME TO`` operation, including the necessary infrastructure for connector implementations. :pr:`23749`

Iceberg Connector Changes
* Add support for ``ALTER VIEW RENAME TO``. :pr:`23749`

Memory Connector Changes
* Add support for ``ALTER VIEW RENAME TO``. :pr:`23749`
```

steveburnett commented:
Consider adding documentation for this to https://prestodb.io/docs/current/connector/iceberg.html#sql-support.

steveburnett reviewed (dismissed):
LGTM! (docs)

Pull branch, local doc build, looks good. Thanks!

agrawalreetika commented:
Thanks for the PR.

I have a few suggestions -
- I think we can add one commit about adding SQL statement support and different commits for support for `ALTER VIEW RENAME` support for each connector support
- Please add tests as well in respective connector support commits
- We can add `ALTER VIEW` documentation along with 1st commit of SQL Statement support here https://prestodb.io/docs/current/sql.html
- Also it would be good to add SQL statement support for `ALTER VIEW [ IF EXISTS ]` as well and its respective tests

hantangwangd commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -1190,6 +1190,12 @@ The table is partitioned by the transformed value of the column::
 
      ALTER TABLE iceberg.web.page_views ADD COLUMN ts timestamp WITH (partitioning = 'hour');
 
+ALTER VIEW
```
```suggestion
ALTER VIEW
^^^^^^^^^^
```

hantangwangd commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -1190,6 +1190,12 @@ The table is partitioned by the transformed value of the column::
 
+ALTER VIEW
+
+Alter view operations to alter the name of an existing view to a new name is supported in the Iceberg connector::
```
Should we clarify that we only supports altering views in Hive and Glue metastores? The same as `CREATE VIEW` above.

hantangwangd commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -1190,6 +1190,12 @@ The table is partitioned by the transformed value of the column::
+
+Alter view operations to alter the name of an existing view to a new name is supported in the Iceberg connector::
+
+     ALTER TABLE iceberg.web.page_views RENAME TO iceberg.web.page_new_views;
```
```suggestion
     ALTER VIEW iceberg.web.page_views RENAME TO iceberg.web.page_new_views;
```

hantangwangd commented on presto-main/src/main/java/com/facebook/presto/execution/RenameViewTask.java:
```diff
@@ -0,0 +1,88 @@
+    private final AccessControl accessControl;
+
+    @Inject
+    public RenameViewTask(Metadata metadata, AccessControl accessControl)
```
Is it necessary to accept `metadata` and `accessControl` here?

hantangwangd commented on presto-main/src/main/java/com/facebook/presto/execution/RenameViewTask.java:
```diff
@@ -0,0 +1,88 @@
+
+    public ListenableFuture<?> execute(RenameView statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, Session session, List<Expression> parameters, WarningCollector warningCollector)
+    {
+        QualifiedObjectName name = createQualifiedObjectName(session, statement, statement.getSource());
+        QualifiedObjectName viewName = createQualifiedObjectName(session, statement, statement.getSource());
```
Why declare two variables for the same view name here?

hantangwangd commented on presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java:
```diff
@@ -566,6 +568,7 @@ private LocalQueryRunner(Session defaultSession, FeaturesConfig featuresConfig,
                 .put(DropMaterializedView.class, new DropMaterializedViewTask())
                 .put(RenameColumn.class, new RenameColumnTask())
                 .put(RenameTable.class, new RenameTableTask())
+                .put(RenameView.class, new RenameViewTask(metadata, accessControl))
```
The same as mentioned above, is it necessary to pass `metadata` and `accessControl` on construction of `RenameViewTask`?

ZacBlanco commented:
@dmariamgeorge Can we split this into two commits like @agrawalreetika suggested? Once we have both commits I will also review

steveburnett commented:
Minor nits of formatting suggested for the release note entry.

```
== RELEASE NOTES ==
General Changes
* Add support for ``ALTER VIEW RENAME TO``. :pr:`23749`

Iceberg Connector Changes
* Add support for ``ALTER VIEW RENAME TO``. :pr:`23749`
```

dmariamgeorge commented:
> @dmariamgeorge Can we split this into two commits like @agrawalreetika suggested? Once we have both commits I will also review

Sure @ZacBlanco. Working on the same.

dmariamgeorge commented:
> > @dmariamgeorge Can we split this into two commits like @agrawalreetika suggested? Once we have both commits I will also review
> 
> Sure @ZacBlanco. Working on the same.

@ZacBlanco - I have split the commits as suggested by @agrawalreetika. Can you please take a look?

agrawalreetika commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergMetadataListing.java:
```diff
@@ -153,4 +153,16 @@ public void testTableValidation()
+    {
+        assertQuerySucceeds("CREATE TABLE iceberg.test_schema.iceberg_test_table (_string VARCHAR, _integer INTEGER)");
+        assertUpdate("CREATE VIEW iceberg.test_schema.test_view_to_be_renamed AS SELECT * FROM iceberg.test_schema.iceberg_test_table");
+        assertUpdate("ALTER VIEW IF EXISTS iceberg.test_schema.test_view_to_be_renamed RENAME TO iceberg.test_schema.test_view_renamed");
```
Can you please add 2 more test here when table doesn't exist. One with `IF EXISTS` and one without it?

agrawalreetika commented on presto-docs/src/main/sphinx/sql/alter-view.rst:
```diff
@@ -0,0 +1,26 @@
+Renaming a view does not affect the data or structure of the underlying
+query used to define the view. Any permissions or dependencies on the
+view are retained, and queries or applications using the old name will
+need to be updated to reference.
```
Lets add an Example section as well here, ref -  https://prestodb.io/docs/current/sql/drop-view.html#examples

agrawalreetika commented on presto-main/src/main/java/com/facebook/presto/execution/RenameViewTask.java:
```diff
@@ -0,0 +1,76 @@
+        Optional<ViewDefinition> view = metadata.getMetadataResolver(session).getView(viewName);
+        if (!view.isPresent()) {
+            if (!statement.isExists()) {
+                throw new SemanticException(MISSING_TABLE, statement, "View '%s' does not exist", viewName);
```
Lets add a new SemanticErrorCode for `MISSING_VIEW`

agrawalreetika commented on presto-main/src/main/java/com/facebook/presto/execution/RenameViewTask.java:
```diff
@@ -0,0 +1,76 @@
+            throw new SemanticException(MISSING_CATALOG, statement, "Target catalog '%s' does not exist", target.getCatalogName());
+        }
+        if (metadata.getMetadataResolver(session).getView(target).isPresent()) {
+            throw new SemanticException(TABLE_ALREADY_EXISTS, statement, "Target view '%s' already exists", target);
```
Lets add a new SemanticErrorCode for `VIEW_ALREADY_EXISTS`

agrawalreetika commented on presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:
```diff
@@ -1504,6 +1505,12 @@ public void testTruncateTable()
+    public void testRenameView()
+    {
+        assertStatement("ALTER VIEW a RENAME TO b", new RenameView(QualifiedName.of("a"), QualifiedName.of("b"), false));
+    }
```
Lets add a test for `IF EXISTS` as well in the Parser

agrawalreetika reviewed (changes requested):
Thank you for making the changes. 
Looks like some of file changes need to be shifted in respective commits. I have dropped in some review comments please check.

agrawalreetika reviewed (dismissed):
@dmariamgeorge Thank you for making the changes. LGTM

steveburnett commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -1190,6 +1190,15 @@ The table is partitioned by the transformed value of the column::
+ALTER VIEW
+^^^^^^^^^^
+
+Alter view operations to alter the name of an existing view to a new name is supported in the Iceberg connector::
```
```suggestion
Alter view operations to alter the name of an existing view to a new name is supported in the Iceberg connector.
```

Using `::` here causes the code block to look like this in a local doc build: 
<img width="984" alt="Screenshot 2024-11-13 at 10 44 22 AM" src="https://github.com/user-attachments/assets/481d3e50-ecac-4155-92ed-8c41ff895a2b">

I tested this edit in a local doc build and it fixes the code block.

steveburnett commented on presto-docs/src/main/sphinx/connector/memory.rst:
```diff
@@ -92,6 +92,15 @@ To delete an existing table:
+ALTER VIEW
+^^^^^^^^^^
+
+Alter view operations to alter the name of an existing view to a new name is supported in the Memory connector::
```
```suggestion
Alter view operations to alter the name of an existing view to a new name is supported in the Memory connector.
```
Same - using `::` here causes the code block to look like this in a local doc build: 

<img width="974" alt="Screenshot 2024-11-13 at 10 48 00 AM" src="https://github.com/user-attachments/assets/97b53a84-9706-4e8b-96a6-392bd9d23796">

I tested this edit in a local doc build and it fixes the code block.

steveburnett commented on presto-docs/src/main/sphinx/sql/alter-view.rst:
```diff
@@ -0,0 +1,42 @@
+============
```
```suggestion
==========
```
Formatting lines should be the same number of characters as the text being formatted.

steveburnett commented on presto-docs/src/main/sphinx/sql/alter-view.rst:
```diff
@@ -0,0 +1,42 @@
+============
+ALTER VIEW
+============
```
```suggestion
==========
```

steveburnett reviewed (changes requested):
Thanks for the doc! Just a few formatting nits and such.

hantangwangd reviewed (changes requested):
Thanks for this work, one little problem and some nits, otherwise looks good to me.

ZacBlanco commented on presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:
```diff
@@ -431,6 +432,24 @@ public void checkCanCreateView(ConnectorTransactionHandle transaction, Connector
+                context.getSource(),
+                Optional.empty(),
+                false,
+                HiveColumnConverterProvider.DEFAULT_COLUMN_CONVERTER_PROVIDER,
```
nit: static import

ZacBlanco commented on presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java:
```diff
@@ -431,6 +432,24 @@ public void checkCanCreateView(ConnectorTransactionHandle transaction, Connector
+    public void checkCanRenameView(ConnectorTransactionHandle transaction, ConnectorIdentity identity, AccessControlContext context, SchemaTableName viewName, SchemaTableName newViewName)
+    {
+        MetastoreContext metastoreContext = new MetastoreContext(
+                identity, context.getQueryId().getId(),
```
nit: split these arguments onto two lines

ZacBlanco commented on presto-hive/src/main/java/com/facebook/presto/hive/security/SystemTableAwareAccessControl.java:
```diff
@@ -166,6 +166,12 @@ public void checkCanCreateView(ConnectorTransactionHandle transactionHandle, Con
+    @Override
+    public void checkCanRenameView(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, AccessControlContext context, SchemaTableName viewName, SchemaTableName newViewName)
+    {
+        delegate().checkCanRenameView(transactionHandle, identity, context, viewName, newViewName);
```
this should be `delegate`, not `delegate()`

ZacBlanco commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergMetadataListing.java:
```diff
@@ -153,4 +153,21 @@ public void testTableValidation()
+    {
+        assertQuerySucceeds("CREATE TABLE iceberg.test_schema.iceberg_test_table (_string VARCHAR, _integer INTEGER)");
+        assertUpdate("CREATE VIEW iceberg.test_schema.test_view_to_be_renamed AS SELECT * FROM iceberg.test_schema.iceberg_test_table");
+        assertUpdate("ALTER VIEW IF EXISTS iceberg.test_schema.test_view_to_be_renamed RENAME TO iceberg.test_schema.test_view_renamed");
```
Can we add a case without the `IF EXISTS` clause?

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/execution/RenameViewTask.java:
```diff
@@ -0,0 +1,76 @@
+            throw new SemanticException(VIEW_ALREADY_EXISTS, statement, "Target view '%s' already exists", target);
+        }
+        if (!viewName.getCatalogName().equals(target.getCatalogName())) {
+            throw new SemanticException(NOT_SUPPORTED, statement, "View rename across catalogs is not supported");
```
We support renaming across schemas though? Does it make sense to rename across schemas? What if the source tables don't exist in the new schema?

ZacBlanco commented on presto-memory/src/test/java/com/facebook/presto/plugin/memory/TestMemorySmoke.java:
```diff
@@ -185,6 +185,31 @@ public void testViews()
     }
 
+    @Test
+    public void testRenameView()
```
Wondering if we should instead add these tests to `AbstractTestDistributedQueries`? This way all connectors which support views can run this and we don't need to duplicate code.

dmariamgeorge replied:
@ZacBlanco - The support for ALTER VIEW RENAME TO is implemented only for the Iceberg and Memory connectors. If we add a test for this in AbstractTestDistributedQueries, it is required to override it in each connector, so the test will not fail for the other connectors.

steveburnett reviewed (dismissed):
LGTM! (docs)

Pull updated branch, new local doc build, reviewed, looks good. Thanks!

hantangwangd commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -511,12 +511,7 @@ public synchronized MetastoreOperationResult renameTable(MetastoreContext metast
         requireNonNull(tableName, "tableName is null");
         requireNonNull(newDatabaseName, "newDatabaseName is null");
         requireNonNull(newTableName, "newTableName is null");
-
-        Table table = getRequiredTable(metastoreContext, databaseName, tableName);
         getRequiredDatabase(metastoreContext, newDatabaseName);
-        if (isIcebergTable(table)) {
-            throw new PrestoException(NOT_SUPPORTED, "Rename not supported for Iceberg tables");
-        }
```
As I understand, it's not just for testing. When user set `hive.metastore=file` and set `hive.metastore.catalog.dir` to a local or hdfs path, they are indeed using `FileHiveMetastore` in their environment. Meanwhile, it's not a good idea to introduce additional problems in any code, even for code that is only used in testing.

Iceberg views created by hive catalog will have a table parameters `presto_view=true`, so I think we can check this parameter, let iceberg view to pass, and keep iceberg table as it is until we really support renaming iceberg table name in hive catalog. Do you think this makes sense?

dmariamgeorge replied:
Sure @hantangwangd.

hantangwangd reviewed (commented):
Thanks for the fix, LGTM. Just one little nit.

hantangwangd commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:
```diff
@@ -1026,6 +1027,11 @@ public static boolean isIcebergTable(Map<String, String> tableParameters)
         return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(tableParameters.get(ICEBERG_TABLE_TYPE_NAME));
     }
 
+    public static boolean isIcebergView(Table table)
+    {
+        return "true".equalsIgnoreCase(table.getParameters().get(ICEBERG_VIEW_TYPE_NAME));
+    }
```
Let's use the already existed constant `PRESTO_VIEW_FLAG`, no need to declare another one.

hantangwangd reviewed (approved):
LGTM, you can rebase to the newest commit list to fix the CI test failure.

steveburnett commented on presto-docs/src/main/sphinx/sql/alter-view.rst:
```diff
@@ -0,0 +1,42 @@
+
+Renaming a view does not affect the data or structure of the underlying
+query used to define the view. Any permissions or dependencies on the
+view are retained, and queries or applications using the old name will
```
```suggestion
view are retained, and queries or applications using the old name must
```

steveburnett commented on presto-docs/src/main/sphinx/sql/alter-view.rst:
```diff
@@ -0,0 +1,42 @@
+Renaming a view does not affect the data or structure of the underlying
+query used to define the view. Any permissions or dependencies on the
+view are retained, and queries or applications using the old name will
+need to be updated to reference.
```
```suggestion
be updated to use the new name.
```

steveburnett reviewed (changes requested):
Thanks for the updated doc! Two minor rephrasing suggestions.

dmariamgeorge commented:
Hi @ZacBlanco @agrawalreetika @steveburnett @hantangwangd 

Could you please take another look at this PR? I’ve incorporated all the review comments. Thanks!

steveburnett reviewed (approved):
LGTM! (docs)

Pull updated build, new local doc build, reviewed pages in local doc build. Everything looks good. Thanks!

ZacBlanco commented:
@dmariamgeorge instead of the screenshot in the PR description, can you please replace it with a short summary of the unit and integration tests that have been added

ZacBlanco commented:
Also, I suggest in the release note section to be more specific in the "General Changes" section that we're adding the ability for connectors to implement `ALTER VIEW RENAME TO`. IMO It isn't clear from the two release notes that it contains the infrastructure and only an implementation for the Iceberg connector.

dmariamgeorge commented:
@ZacBlanco - I have revised the PR description. Could you please review it?

dmariamgeorge commented:
Hi @tdcmeehan 

I have received approvals from @ZacBlanco, @agrawalreetika, @hantangwangd, and @steveburnett on this PR. Could you please help merge this PR? Thanks!
